Organize F# language guide toc/index, add F# scenarios (#26804)
dotnet/docs: docs/fsharp/language-reference/index.md

## Values and Functions

The following table shows reference topics available that describe language concepts related to values, `let`-bindings and functions.

|Title|Description|
|-----|-----------|
|[Values](./values/index.md)|Learn about values, which are immutable quantities that have a specific type; values can be integral or floating point numbers, characters or text, lists, sequences, arrays, tuples, discriminated unions, records, class types, or function values.|
|[Functions](./functions/index.md)|Functions are the fundamental unit of program execution in any programming language. An F# function has a name, can have parameters and take arguments, and has a body. F# also supports functional programming constructs such as treating functions as values, using unnamed functions in expressions, composition of functions to form new functions, curried functions, and the implicit definition of functions by way of the partial application of function arguments.|
|[Function Expressions](./functions/lambda-expressions-the-fun-keyword.md)|Learn how to use the F# 'fun' keyword to define a lambda expression, which is an anonymous function.|

## Loops and Conditionals

> The following table lists topics that describe F# loops and conditionals.

Reviewer: ```suggestion
The following table lists articles that describe F# loops and conditionals.
```

|Title|Description|
|-----|-----------|
|[Conditional Expressions: `if...then...else`](conditional-expressions-if-then-else.md)|Learn about the `if...then...else` expression, which runs different branches of code and also evaluates to a different value depending on the Boolean expression given.|
|[Loops: `for...in` Expression](loops-for-in-expression.md)|Learn about the `for...in` expression, a looping construct that is used to iterate over the matches of a pattern in an enumerable collection such as a range expression, sequence, list, array, or other construct that supports enumeration.|
|[Loops: `for...to` Expression](loops-for-to-expression.md)|Learn about the `for...to` expression, which is used to iterate in a loop over a range of values of a loop variable.|
|[Loops: `while...do` Expression](loops-while-do-expression.md)|Learn about the `while...do` expression, which is used to perform iterative execution (looping) while a specified test condition is true.|

## Pattern Matching

> The following table shows reference topics available that describe language concepts.

Reviewer: ```suggestion
The following table shows reference articles that describe language concepts.
```

|Title|Description|
|-----|-----------|
|[Pattern Matching](pattern-matching.md)|Learn about patterns, which are rules for transforming input data and are used throughout the F# language. You can compare data with a pattern, decompose data into constituent parts, or extract information from data in various ways.|
|[Match Expressions](match-expressions.md)|Learn about the `match` expression, which provides branching control that is based on the comparison of an expression with a set of patterns.|
|[Active Patterns](active-patterns.md)|Learn about active patterns. Active patterns enable you to define named partitions that subdivide input data. You can use active patterns to decompose data in a customized manner for each partition.|

## Exception Handling

> The following table shows reference topics available that describe language concepts related to exception handling.

Reviewer: ```suggestion
The following table shows reference articles that describe language concepts related to exception handling.
```

|Title|Description|
|-----|-----------|
|[Exception Handling](./exception-handling/index.md)|Contains information about exception handling support in F#.|
|[The `try...with` Expression](exception-handling/the-try-with-expression.md)|Learn about how to use the `try...with` expression for exception handling.|
|[The `try...finally` Expression](exception-handling/the-try-finally-expression.md)|Learn about how the F# `try...finally` expression enables you to execute clean-up code even if a block of code throws an exception.|

> |[The `use` Keyword](resource-management-the-use-keyword.md)|Learn about the keywords `use` and `using`, which can control the initialization and release of resources|

Reviewer: ```suggestion
|[The `use` Keyword](resource-management-the-use-keyword.md)|Learn about the keywords `use` and `using`, which can control the initialization and release of resources.|
```

|[Assertions](assertions.md)|Learn about the `assert` expression, which is a debugging feature that you can use to test an expression. Upon failure in Debug mode, an assertion generates a system error dialog box.|

## Types and Type Inference

> The following table shows reference topics available that describe how types and type inference work in F#.

Reviewer: ```suggestion
The following table shows reference articles that describe how types and type inference work in F#.
```

|Title|Description|
|-----|-----------|
|[Types](fsharp-types.md)|Learn about the types that are used in F# and how F# types are named and described.|

> |[Basic Types](basic-types.md)|Learn about the fundamental basic types that are used in F#. It also provides the corresponding .NET types and the minimum and maximum values for each type.|

Reviewer: ```suggestion
|[Basic Types](basic-types.md)|Learn about the fundamental types that are used in F#. It also provides the corresponding .NET types and the minimum and maximum values for each type.|
```

|[Generics](./generics/index.md)|Learn about generic constructs in F#.|
|[Automatic Generalization](./generics/automatic-generalization.md)|Learn about how F# automatically generalizes the arguments and types of functions so that they work with multiple types when possible.|
|[Constraints](casting-and-conversions.md)|Learn about constraints that apply to generic type parameters to specify the requirements for a type argument in a generic type or function.|
|[Flexible Types](flexible-types.md)|Learn about flexible types. A flexible type annotation is an indication that a parameter, variable, or value has a type that is compatible with type specified, where compatibility is determined by position in an object-oriented hierarchy of classes or interfaces.|
|[Units of Measure](units-of-measure.md)|Learn about units of measure. Floating point values in F# can have associated units of measure, which are typically used to indicate length, volume, mass, and so on.|
|[Byrefs](byrefs.md)|Learn about byref and byref-like types in F#, which are used for low-level programming.|

## Tuples, Lists, Collections, Options

> The following table shows reference topics available that describe types supported by F#.

Reviewer: ```suggestion
The following table shows reference articles that describe types supported by F#.
```

|[Collections](fsharp-collection-types.md)|An overview of the F# functional collection types, including types for arrays, lists, sequences (seq), maps, and sets.|
|[Lists](lists.md)|Learn about lists. A list in F# is an ordered, immutable series of elements all of the same type.|
|[Options](options.md)|Learn about the option type. An option in F# is used when a value may or may not exist. An option has an underlying type and may either hold a value of that type or it may not have a value.|
|[Arrays](arrays.md)|Learn about arrays. Arrays are fixed-size, zero-based, mutable sequences of consecutive data elements, all of the same type.|
|[Sequences](sequences.md)|Learn about sequences. A sequence is a logical series of elements all of one type. Individual sequence elements are only computed if necessary, so the representation may be smaller than a literal element count indicates.|
|[Reference Cells](reference-cells.md)|Learn about reference cells, which are storage locations that enable you to create mutable variables with reference semantics.|

## Records and Discriminated Unions

> The following table shows reference topics available that describe record and discriminated union type definitions supported by F#.

Reviewer: ```suggestion
The following table shows reference articles that describe record and discriminated union type definitions supported by F#.
```

|Title|Description|
|-----|-----------|
|[Records](records.md)|Learn about records. Records represent simple aggregates of named values, optionally with members.|

> |[Anonymous Records](anonymous-records.md)|Learn how to use construct and use anonymous records, a language feature that helps with the manipulation of data.|

Reviewer: ```suggestion
|[Anonymous Records](anonymous-records.md)|Learn how to construct and use anonymous records, a language feature that helps with the manipulation of data.|
```

|[Discriminated Unions](discriminated-unions.md)|Learn about discriminated unions, which provide support for values that may be one of a variety of named cases, each with possibly different values and types.|
|[Structs](structures.md)|Learn about structs, which are compact object types that can be more efficient than a class for types that have a small amount of data and simple behavior.|

> |[Enumerations](enumerations.md)|Learn about enumerations are types that have a defined set of named values. You can use them in place of literals to make code more readable and maintainable.|

Reviewer: ```suggestion
|[Enumerations](enumerations.md)|Enumerations are types that have a defined set of named values. You can use them in place of literals to make code more readable and maintainable.|
```

## Object Programming

> The following table shows reference topics available that describe F# object programming.

Reviewer: ```suggestion
The following table shows reference articles that describe F# object programming.
```

|[Computation Expressions](computation-expressions.md)|Learn about computation expressions in F#, which provide a convenient syntax for writing computations that can be sequenced and combined using control flow constructs and bindings. They can be used to manage data, control, and side effects in functional programs.|
|[Sequence Expressions](sequences.md)|Learn about sequence expressions, which let you generate sequences of data on-demand.|
|[Async Expressions](async-expressions.md)|Learn about async expressions, which let you write asynchronous code in a way that is very close to the way you would naturally write synchronous code.|
|[Task Expressions](task-expressions.md)|Learn about task expressions, which are an alternative way of writing asynchronous code used when interoperating with .NET code that consumes or produces .NET tasks.|
|[Query Expressions](query-expressions.md)|Learn about query expressions, a language feature that implements LINQ for F# and enables you to write queries against a data source or enumerable collection.|

## Attributes, Reflection, Quotations and Plain Text Formatting

> The following table lists topics that describe F# reflective features, including attributes, quotations, `nameof` and plain text formatting.

Reviewer: ```suggestion
The following table lists articles that describe F# reflective features, including attributes, quotations, `nameof`, and plain text formatting.
```

|Title|Description|
|-----|-----------|
|[Attributes](attributes.md)|Learn how F# Attributes enable metadata to be applied to a programming construct.|
|[nameof](nameof.md)|Learn about the `nameof` operator, a metaprogramming feature that allows you to produce the name of any symbol in your source code.|
|[Caller Information](caller-information.md)|Learn about how to use Caller Info Argument Attributes to obtain caller information from a method.|

> |[Source Line, File, and Path Identifiers](source-line-file-path-identifiers.md)|Learn about the identifiers `__LINE__`, `__SOURCE_DIRECTORY__`, and `__SOURCE_FILE__`, which are built-in values that enable you to access the source line number, directory and file name in your code.|

Reviewer: ```suggestion
|[Source Line, File, and Path Identifiers](source-line-file-path-identifiers.md)|Learn about the identifiers `__LINE__`, `__SOURCE_DIRECTORY__`, and `__SOURCE_FILE__`, which are built-in values that enable you to access the source line number, directory, and file name in your code.|
```

|[Code Quotations](code-quotations.md)|Learn about code quotations, a language feature that enables you to generate and work with F# code expressions programmatically.|
|[Plain Text Formatting](plaintext-formatting.md)|Learn how to use sprintf and other plain text formatting in F# applications and scripts.|

## Type Providers

> The following table lists topics that describe F# type providers.

Reviewer: ```suggestion
The following table lists articles that describe F# type providers.
```

|Title|Description|
|-----|-----------|

> |[Type Providers](../tutorials/type-providers/index.md)|Learn about type provides and provides links to walkthroughs on using the built-in type providers to access databases and web services.|

Reviewer: ```suggestion
|[Type Providers](../tutorials/type-providers/index.md)|Learn about type providers and find links to walkthroughs on using the built-in type providers to access databases and web services.|
```

> |[Create a Type Provider](../tutorials/type-providers/creating-a-type-provider.md)|Learn how to create your own F# type providers by examining several simple type providers to illustrate the basic concepts.|

Reviewer: ```suggestion
|[Create a Type Provider](../tutorials/type-providers/creating-a-type-provider.md)|Learn how to create your own F# type providers by examining several simple type providers that illustrate the basic concepts.|
```
